WFLY-14133 Add call-timeout attribute to JMS core bridge

TomasHofman opened the pull request:
Issue: https://issues.redhat.com/browse/WFLY-14133

The call timeout attribute is not currently configurable via the management model. This PR makes it accessible.

Work in progress.

marekkopecky commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BridgeAdd.java:
```diff
@@ -173,41 +175,55 @@ static BridgeConfiguration createBridgeConfiguration(final OperationContext cont
+
         try {
-            String transformerClassName = bridgeConfig.getTransformerConfiguration() != null ? bridgeConfig.getTransformerConfiguration().getClassName() : null;
-            if (bridgeConfig.getDiscoveryGroupName() != null) {
```
Why is this bridgeConfig settings no longer needed because of adding "call-timeout" attribute?

TomasHofman replied:
I'm gonna close this PR for now, this was the version from before it was recommended to modify Artemis API first.

TomasHofman replied:
The BridgeConfig instance will now be directly passed to `ActiveMQServer#deployBridge(BridgeConfig)`, rather than the attributes being unpacked from BridgeConfig and passed to `ActiveMQServerControl#createBridge(String, String, String, ...)` explicitely.

(In the original impl, the `ActiveMQServerControlImpl#createBridge(String, String, String, ...)` then reconstructs the `BridgeConfig` instance and passes it to `ActiveMQServer#deployBridge(BridgeConfig)` anyway, so we are just taking a shortcut here.)

Not sure if this was what you were asking though :).

marekkopecky replied:
Thanks!

TomasHofman commented:
@marekkopecky @ehsavoie I'm reopening this PR, as it needs to be implemented purely on the Wildfly side, since Atemis 2.17 will not be available for some time yet.

TomasHofman commented:
Going to update the test plan and analysis doc shortly.

marekkopecky commented:
> An integration test will be added that verifies creation of a bridge and verifies the created bridge transfers messages.

@TomasHofman I'm not seeing integration test here, can you please point me to that test (if it's already exists) or add this test to this PR?

TomasHofman commented:
> > An integration test will be added that verifies creation of a bridge and verifies the created bridge transfers messages.
> 
> @TomasHofman I'm not seeing integration test here, can you please point me to that test (if it's already exists) or add this test to this PR?

I didn't finish it yet, I will add it to this PR and let you know.

TomasHofman committed:
WFLY-14133 Add call-timeout attribute to JMS core bridge

TomasHofman commented:
@marekkopecky @ehsavoie I rebased on the existing PR with messaging subsystem schema version bump, and added setting from system property.

An integration test that can serve as a regression test already exists in the messaging-testsuite:https://gitlab.hosts.mwqe.eng.bos.redhat.com/jbossqe-eap/messaging-testsuite/-/blob/master/jboss-hornetq-testsuite/tests-eap7/src/test/java/org/jboss/qa/artemis/test/bridges/TransferOverBridgeTestCase.java

TomasHofman commented:
@ehsavoie I added an integration test to manual mode so that we add it to one module only. It's passing locally, I'm waiting for CI to run...

TomasHofman commented:
@marekkopecky Emmanuel would like to dump the `BridgeAddTestCase` (or at least all the `testPerformRuntime*` methods in it) and only keep the integration test and the subsystem tests. If you are OK with that I will post another PR to test plan...

ehsavoie commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BridgeAdd.java:
```diff
@@ -152,6 +162,10 @@ static BridgeConfiguration createBridgeConfiguration(final OperationContext cont
                 .setUser(user)
                 .setPassword(password);
 
+        if (callTimeout != null) {
```
shoudn't be the case since you have a default value

TomasHofman replied:
Check removed.

ehsavoie commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BridgeDefinition.java:
```diff
@@ -185,7 +185,8 @@
             USER, PASSWORD, CREDENTIAL_REFERENCE,
-            CONNECTOR_REFS, DISCOVERY_GROUP_NAME
+            CONNECTOR_REFS, DISCOVERY_GROUP_NAME,
+            CommonAttributes.CALL_TIMEOUT
```
I'm wondering if the bridge should have its own attribute definition

TomasHofman replied:
OK, created it's own attr definition.

ehsavoie commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingTransformerRegistration.java:
```diff
@@ -83,10 +84,20 @@ public void registerTransformers(SubsystemTransformerRegistration registration)
+        ResourceTransformationDescriptionBuilder server = subsystem.addChildResource(MessagingExtension.SERVER_PATH);
+
+        ResourceTransformationDescriptionBuilder bridge = server.addChildResource(MessagingExtension.BRIDGE_PATH);
+        bridge.getAttributeBuilder()
```
you could use  rejectDefinedAttributeWithDefaultValue for this

TomasHofman replied:
Done.

ehsavoie commented on messaging-activemq/src/test/java/org/wildfly/extension/messaging/activemq/BridgeAddTestCase.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    @Test
+    public void testPerformRuntime() throws Exception {
```
Doesn't have much value and is covered by the integration testing

TomasHofman replied:
File removed.

ehsavoie commented on messaging-activemq/src/test/java/org/wildfly/extension/messaging/activemq/BridgeAddTestCase.java:
```diff
@@ -0,0 +1,210 @@
+    /**
+     * If the ActiveMQ server is not started, IllegalStateException is thrown.
+     */
+    @Test(expected = IllegalStateException.class)
```
Doesn't have much value and is covered by the integration testing

TomasHofman replied:
Files removed.

ehsavoie commented on messaging-activemq/src/test/java/org/wildfly/extension/messaging/activemq/BridgeAddTestCase.java:
```diff
@@ -0,0 +1,210 @@
+     * during subsystem initialization instead...)
+     */
+    @Test
+    public void testPerformRuntime_subsystemNotInitialized() throws Exception {
```
Doesn't have much value and is covered by the integration testing

TomasHofman replied:
I removed the test file altogether.

ehsavoie commented on messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties:
```diff
@@ -43,6 +43,7 @@ address-setting.slow-consumer-threshold=The minimum rate of message consumption
 address-setting=An address setting defines some attributes that are defined against an address wildcard rather than a specific queue.
 bindings-directory.path=The directory in which the bindings journal lives. The default is ${jboss.server.data.dir}/messagingbindings.
 bridge.add=Operation adding a bridge.
+bridge.call-timeout=Timeout (in milliseconds) for blocking calls.
```
Don't specify the unit there as it is already defined in the attribute definition

TomasHofman replied:
OK...

ehsavoie commented on messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_13_0.xsd:
```diff
@@ -0,0 +1,1043 @@
+                    <xs:attribute name="password" type="xs:string"  use="optional" />
+                    <xs:attribute name="static-connectors" type="stringList"  use="optional" />
+                    <xs:attribute name="discovery-group" type="xs:string"  use="optional" />
+                    <xs:attribute name="call-timeout" type="xs:long"  use="optional" />
```
maybe you should set the default value there

TomasHofman replied:
Done.

ehsavoie commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/messaging/CoreBridgeCallTimeoutTestCase.java:
```diff
@@ -0,0 +1,220 @@
+        // clean up
+        jmsOperations.removeJmsQueue(SOURCE_QUEUE_NAME);
+        jmsOperations.removeJmsQueue(TARGET_QUEUE_NAME);
+        jmsOperations.removeCoreBridge(BRIDGE_NAME);
```
you need to close the management client

ehsavoie replied:
and the jmsOperations (even if this does nothing for now)

TomasHofman replied:
Closed both in `tearDown()`.

ehsavoie commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/messaging/CoreBridgeCallTimeoutTestCase.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    @Test(timeout = 300000)
+    public void testCallTimeout() throws Exception {
```
Could you split this test into 2 tests: one in standard test case and one with the failure ?

TomasHofman replied:
Done.

TomasHofman commented:
Updated - I removed the unit tests, there are integ test and subsystem tests now.

I think there was a legitimate CI failure under security manager, which I tried to fix. I will see if it passes now.

marekkopecky commented:
> Emmanuel would like to dump the BridgeAddTestCase (or at least all the testPerformRuntime* methods in it) and only keep the integration test and the subsystem tests. If you are OK with that I will post another PR to test plan...

Yes, I'm ok with it.

marekkopecky commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/messaging/CoreBridgeCallTimeoutTestCase.java:
```diff
@@ -0,0 +1,242 @@
+        jmsOperations.removeJmsQueue(SOURCE_QUEUE_NAME);
+        jmsOperations.removeJmsQueue(TARGET_QUEUE_NAME);
+        jmsOperations.removeCoreBridge(BRIDGE_NAME);
+        container.stop(CONTAINER_QUALIFIER);
```
Why we need to stop the container after each test? What about `s/@After/@AfterClass/` and the same for setUp method? Btw, why we actually need to have this test in manual mode?

marekkopecky replied:
Ah, manualmode is needed because of byteman arquillian container. But still I hope that we could remove one reload by `@After` -> `@AfterClass` and related changes

TomasHofman replied:
I not happy with the manualmode module either. The reason I put it there is that Emmanuel already has another PR with byteman integration that will go into the manualmode module.

I don't think it's desirable to have byteman in more modules than is necessary. But if you think otherwise I can move it.

I will look into the restarts, agree on that.

TomasHofman replied:
@marekkopecky I implemented it via test method ordering, PR is updated.

TomasHofman commented:
For the record, all CI jobs passed except "Windows - JDK 11" one which timed out.

marekkopecky commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BridgeAdd.java:
```diff
@@ -65,6 +68,8 @@
 
     public static final BridgeAdd INSTANCE = new BridgeAdd();
 
+    static final String CALL_TIMEOUT_PROPERTY = "org.wildfly.messaging.core.bridge.call-timeout";
```
Do we have a test for this property?

TomasHofman replied:
I returned the unit test back in.

TomasHofman commented:
@ehsavoie @marekkopecky I updated the integration test to use `WildflyTestRunner` instead of Arquillian. That way I can use `@BeforeClass`/`@AfterClass` to start and stop container and don't need to use test counter.

Emmanuel, if it looks alright to you, Marek could start their testsuite runs and start working on precheck.

TomasHofman commented:
Good morning @marekkopecky @ehsavoie, PR is rebased.

ehsavoie reviewed (approved):
Please squash the commits

ehsavoie commented:
LGTM :+1:

darranl commented on messaging-activemq/src/test/java/org/wildfly/extension/messaging/activemq/BridgeAddTestCase.java:
```diff
@@ -0,0 +1,67 @@
+package org.wildfly.extension.messaging.activemq;
```
Missing copyright header.

TomasHofman replied:
Thanks for noticing, updated.

darranl commented on testsuite/integration/manualmode/src/test/java/org/jboss/as/test/manualmode/messaging/CoreBridgeCallTimeoutTestCase.java:
```diff
@@ -0,0 +1,274 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2012, Red Hat, Inc., and individual contributors
```
2021

marekkopecky commented:
@TomasHofman Can you please resolve conflicts?

TomasHofman commented:
@marekkopecky Sure, rebased.

bstansberry commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BridgeAdd.java:
```diff
@@ -166,48 +179,77 @@ static BridgeConfiguration createBridgeConfiguration(final OperationContext cont
+        // extracted from ActiveMQServerControlImpl#checkStarted()
+
+        if (!server.isStarted()) {
+            throw new IllegalStateException("Broker is not started. It can not be managed yet");
```
@TomasHofman this needs to be an i18n message

bstansberry replied:
also s/can not/cannot

TomasHofman replied:
Thanks @bstansberry , fixed.

bstansberry reviewed (changes requested):
@ehsavoie Please review
